### DISCUSSION TITLE
How can I get the partial derivative of the transformation matrix w.r.t. the joint angles

### DISCUSSION THREAD

**Discussion by BolunDai0216** [stack-of-tasks/pinocchio#1939] [Q&A]
Hi,

I was wondering, for a robot manipulator, is there an easy way to get the derivative of a transformation matrix of a link with respect to the joint angles, i.e., $\frac{\partial T}{\partial q}$, in pinocchio?

Thanks in advance!

**Comment by fabinsch** [comment]
Hi @BolunDai0216,

what you want to compute is `J(q) = dT(q) / dq` which is the regular jacobian. In Pinocchio, you can use `computeForwardKinematicsDerivatives` for this. This function is similar to doing a `forwardKinematics(model,data,q,v)` followed by a `computeJointJacobians(model,data,q)`. Afterward, you can get the Jacobian for the specific joint by calling the `getJointJacobian` function.

**Reply by BolunDai0216** [reply]
@fabinsch Thanks for the response! I was running the following code

```python
# robot is a pinocchio.robot_wrapper instance
q_normalized = pin.normalize(robot.model, q)
ddq = pin.aba(robot.model, robot.data, q_normalized, dq, torque)
pin.computeForwardKinematicsDerivatives(robot.model, robot.data, q_normalized, dq, ddq)
jacobian= robot.getFrameJacobian(self.FRAME_ID, self.jacobian_frame)
```

And I noticed that `jacobian` has shape `(6, 9)`, but for `dT(q) / dq` it should have shape `(4, 4, 9)`. Is this the expected behavior? I get the same shape after calling `getJointJacobian`.

I added the `pin.normalize` after reading #1534, is this necessary?

**Reply by BolunDai0216** [reply]
Are you aware of any built-in functions that map a twist $\mathcal{V} = [v, \omega]^T\in\mathbb{R}^6$ into its matrix representation $[\mathcal{V}] = [[[\omega], v], [0, 0]]$?

**Reply by fabinsch** [reply]
I think you can use `pin.exp6` for this. See the docstring here

```
pin.exp6?
Docstring:
exp6( (Motion)motion) -> SE3 :
    Exp: se3 -> SE3. Return the integral of the input spatial velocity during time 1.

exp6( (numpy.ndarray)v) -> SE3 :
    Exp: se3 -> SE3. Return the integral of the input spatial velocity during time 1.
Type:      function
```

**Reply by BolunDai0216** [reply]
I think `pin.exp6` outputs the SE(3) transformation matrix, I was wondering if there are any functions that gives the matrix representation for se(3).

And for the partial derivative of the transformation matrix, I don't think `computeForwardKinematicsDerivatives` is giving me the correct dimension, is there something else that needs to be done?

**Reply by fabinsch** [reply]
I'm not sure to really get your point. Can you please provide a minimal but runnable example to show what you are trying to do? I I guessed that your robot has 9 joints, you ask for the Jacobian for a specific frame, so you should get something of dimension (6, 9) as spatial velocities/differences are represented in the tangent space in Pinocchio.

**Reply by BolunDai0216** [reply]
Sure, I can come up with a minimal example.

My point is, for a transformation matrix `T` that is 4x4, if I want to take the partial derivative of `T` with respect to the joint angles, it would be 4x4x9, which is the result of taking the partial derivative of a 4x4 matrix w.r.t. a 9-dimensional vector. 

I understand that the Jacobian `pinocchio` outputs when calling `getFrameJacobian` is 6x9, however, that is not what I was looking for.

**Reply by fabinsch** [reply]
Ok, I understand what you are looking for. This Euclidian way of describing the Jacobian is not the default one, Pinocchio has no built-in function to directly give you the mapping from 6x9 to 4x4x9. But there are all the functions, that you can assemble this quantity by yourself. You can create `pin.Motion` objects from the cols of the jacobian matrix (`twist = pin.Motion(jacobian[:, i])`) and then call methods on it to obtain different representations (`twist.toActionMatrix()`). Then using the transformation for each joint you can obtain the quantity you're looking for.

**Reply by BolunDai0216** [reply]
Thanks for the response. One question I have is with `twist.toActionMatrix()`, after running `twist = pin.Motion(jacobian[:, i])` I don't think `twist` has the member function `.toActionMatrix()`

```console
>>> dir(twist)
['Random', 'Zero', '__add__', '__array__', '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getinitargs__', '__gt__', '__hash__', '__iadd__', '__init__', '__init_subclass__', '__isub__', '__le__', '__lt__', '__module__', '__mul__', '__ne__', '__neg__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__rmul__', '__safe_for_unpickling__', '__setattr__', '__sizeof__', '__str__', '__sub__', '__subclasshook__', '__truediv__', '__weakref__', '__xor__', 'action', 'angular', 'copy', 'cross', 'dualAction', 'isApprox', 'isZero', 'linear', 'np', 'se3Action', 'se3ActionInverse', 'setRandom', 'setZero', 'vector']
```

If you are referring to `twist.action`, I think the output is $[[R, [p]R], [0, R]]$ which 6x6. From my derivation, I think I would need to get $[[[\omega], v], [0, 0]]$ from `twist` and multiply it with the transformation matrix, is there a way to get the $[[[\omega], v], [0, 0]]$ matrix (which is 4x4) using any built-in function. This is not a difficult thing to write by myself, but using a built-in function makes the codebase cleaner.

**Reply by fabinsch** [reply]
you are right. I think the function you are looking for is `toHomogeneousMatrix` which is binded with `homogeneous`. Please check out the se3-implementation of Pinocchio to see all available functions.

**Reply by BolunDai0216** [reply]
I think homogeneous matrices are the same thing as transformation matrices. I took a look at the se3 implementation and I don't think such a functionality exists. I believe $[[[\omega], v], [0, 0]]$ is a fairly common value that might be used, would it be possible to add a function to compute this in future releases?

**Reply by stephane-caron** [reply]
> I think homogeneous matrices are the same thing as transformation matrices.

Homogeneous representations are more general: transformation matrices are the homogeneous representation for $SE(3)$, and there is a homogeneous representation for $se(3)$ as well. You can find more about this in *e.g.* Chapter 2 *Rigid Body Motion* of [A Mathematical Introduction to Robotic Manipulation](http://www.cse.lehigh.edu/~trink/Courses/RoboticsII/reading/murray-li-sastry-94-complete.pdf). Here is a relevant equation from p. 39:

![image](https://user-images.githubusercontent.com/1189580/235463998-ded41679-2a49-4e55-ab2b-180374d12bac.png)

> I took a look at the se3 implementation and I don't think such a functionality exists.

True, `toHomogeneousMatrix` is a function in ``SE3``, while here we are in $se(3)$, which is called ``Motion`` in Pinocchio :sweat_smile: I couldn't find such a function either, and I agree it would be a consistent feature to add. Let's open a [feature request](https://github.com/stack-of-tasks/pinocchio/issues/1945).

In the meantime we can do it in Python:

```python
def to_homogeneous(twist: pin.Motion) -> np.ndarray: 
     homogeneous = np.zeros((4, 4)) 
     homogeneous[:3, :3] = twist.action[:3, :3] 
     homogeneous[:3, 3] = twist.linear 
     return homogeneous
```
